File: MatchModule_New/AI/States/Defence/InterruptionState.cs
/********************************************************************************
 * 文件名：InterruptionState
 * 创建人：
 * 创建时间：2009-12-30 13:14:09
 * 版本：1.0
 * 本文件版本号：1.0
 * 最后更新：
 * 功能说明：
 * 历史修改记录：
 * <author>  <time>           <version >   <desc>
 *********************************************************************************/
using Games.NB.Match.Base.Attributes;
using Games.NB.Match.Base.Interface;

namespace Games.NB.Match.AI.States.Defence
{
    /// <summary>
    /// Represents the interruption state.
    /// </summary>
    [Singleton]
    public sealed class InterruptionState : DefenceState
    {

        /// <summary>
        /// Initialize current <see cref="IState"/>.
        /// </summary>
        public override void Initialize()
        {
            this.StateChain.Add(DefenceState.Instance);
            this.StateCondition.Add(DefenceState.Instance, ValidateInterruptionToDefence);
        }

        /// <summary>
        /// Return the <see cref="IState"/>'s name.
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return "InterruptionState";
        }

        /// <summary>
        /// Represents the instance of current singleton class.
        /// </summary>
        public new static InterruptionState Instance
        {
            get { return _instance; }
        }

        /// <summary>
        /// Enters the <see cref="InterruptionState"/>.
        /// </summary>
        /// <param name="player"></param>
        public override void Enter(IPlayer player)
        {
        }

        /// <summary>
        /// Player to interrupt ball.
        /// </summary>
        /// <param name="player"><see cref="IPlayer"/></param>
        public override void Action(IPlayer player)
        {
            player.InterruptionBall();
        }

        /// <summary>
        /// Decides the nex <see cref="IState"/>
        /// </summary>
        /// <param name="player">Represents the current <see cref="IPlayer"/>.</param>
        /// <param name="preview">Represents the preview <see cref="IState"/></param>
        /// <returns></returns>
        public override IState QuickDecide(IPlayer player, IState preview)
        {
            if (player.Status.Hasball)
            {
                return HoldBallState.Instance;
            }
            else
            {
                return OffBallState.Instance;
            }
        }

        #region encapsulation

        private static readonly InterruptionState _instance = new InterruptionState();

        private InterruptionState()
        {
            this.Stopable = true;
        }

        private static bool ValidateInterruptionToDefence(IPlayer player, IState preview)
        {
            return true;
        }

        #endregion
    }
}

File: SVN球场AI/HB_SNS/AI/States/Defence/InterruptionState.cpp
/********************************************************************************
 * 文件名：InterruptionState.cpp
 * 创建人：宋斌
 * 创建时间：2011-6-10
 * 版本：1.0
 * 本文件版本号：1.0
 * 最后更新：
 * 功能说明：
 * 历史修改记录：
 * <author>  <time>           <version >   <desc>
 *********************************************************************************/
#include "InterruptionState.h"

#include "../HoldBallState.h"
#include "../OffBallState.h"

InterruptionState::InterruptionState()
{
    m_Stopable = true;
}

InterruptionState* InterruptionState::Instance()
{
    static InterruptionState instance;

    return &instance;
}

void InterruptionState::Initialize()
{
    m_StateChain.push_back(DefenceState::Instance());

    m_StateCondition[DefenceState::Instance()]      = ValidateInterruptionToDefence;
}

string InterruptionState::ToString()
{
    return "InterruptionState";
}

void InterruptionState::Enter(IPlayer* player)
{
}

void InterruptionState::Action(IPlayer* player)
{
    player->InterruptionBall();
}

IState* InterruptionState::QuickDecide(IPlayer* player, IState* preview)
{
    if (player->Status()->Hasball())
    {
        return HoldBallState::Instance();
    }
    else
    {
        return OffBallState::Instance();
    }
}

bool InterruptionState::ValidateInterruptionToDefence(IPlayer* player, IState* preview)
{
    return true;
}

bool InterruptionState::IsInState(IState* st)
{
    if (typeid(*this) == typeid(*st)) 
    {
        return true;
    }

    if (typeid(DefenceState) == typeid(*st))
    {
        return true;
    }

    return false;
}
